File: python/examples/04_spread_scanner.py
"""Cross-exchange funding spread for one base asset, sorted by funding rate."""

from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from kairo_funding import ClientLogicError, FundingClient, cli_entry  # noqa: E402


def annualized_pct(rate: float, interval_hours: int) -> float:
    return rate * (24 / interval_hours) * 365 * 100


def main() -> int:
    base = os.environ.get("KAIRO_BASE", "BTC")
    with FundingClient() as client:
        snapshot = client.get_snapshot(base=base)
    rows = FundingClient.parse_compact_rows(snapshot.get("data", []))
    if not rows:
        raise ClientLogicError(f"no rows for base={base}")
    rows.sort(key=lambda r: (r["funding_rate"], r["exchange"]))
    for row in rows:
        ann = annualized_pct(row["funding_rate"], row["funding_interval_hours"])
        print(
            f"{row['exchange']}  rate={row['funding_rate']}  "
            f"ann={ann:.4f}%  interval={row['funding_interval_hours']}h"
        )
    min_row = rows[0]
    max_row = rows[-1]
    spread = max_row["funding_rate"] - min_row["funding_rate"]
    print(
        f"spread = {spread} (max {max_row['exchange']} @ {max_row['funding_rate']}, "
        f"min {min_row['exchange']} @ {min_row['funding_rate']})"
    )
    return 0


if __name__ == "__main__":
    cli_entry(main)

File: typescript/src/examples/04_spread_scanner.ts
/**
 * 04_spread_scanner: cross-exchange spread for a single base asset, sorted
 * by funding rate ascending, plus a summary line.
 */

import { FundingClient } from "../client.js";
import { cliEntry } from "../runner.js";
import type { FundingEntry } from "../types.js";

function annualizedPct(row: FundingEntry): number {
  return row.fundingRate * (24 / row.fundingIntervalHours) * 365 * 100;
}

async function main(): Promise<number> {
  const base = process.env.KAIRO_BASE ?? "BTC";
  const client = new FundingClient();
  const snap = await client.getSnapshot({ base });

  if (snap.count === 0 || snap.data.length === 0) {
    process.stdout.write(`no rows for base=${base}\n`);
    return 4;
  }

  const sorted = [...snap.data].sort(
    (a, b) => a.fundingRate - b.fundingRate || a.exchange.localeCompare(b.exchange),
  );
  for (const row of sorted) {
    process.stdout.write(
      `${row.exchange}  rate=${row.fundingRate}  ann=${annualizedPct(row).toFixed(4)}%  ` +
        `interval=${row.fundingIntervalHours}h\n`,
    );
  }

  const min = sorted[0]!;
  const max = sorted[sorted.length - 1]!;
  const spread = max.fundingRate - min.fundingRate;
  process.stdout.write(
    `spread = ${spread} (max ${max.exchange} @ ${max.fundingRate}, ` +
      `min ${min.exchange} @ ${min.fundingRate})\n`,
  );
  return 0;
}

cliEntry(main);
